### ISSUE TITLE
[Feature Proposal] - Enable auto-creation of new projects

### ISSUE THREAD

**Issue by raschmitt** [gsoft-inc/azure-pipelines-dependency-track#11]
I just found your task today and I think it's great.

In my personal opinion in a `DevOps` process we should not only have a track record of dependencies that devs can lookup over time but to be able to aware teams at the time of adding code that they might have something bad under the hood. Because anyone can agree that, it's way easier to deal with this type of issue when you have just added packages to the code than six months later when you have a hole monolith that depends on it. And your solution does it perfectly, with thresholds, cheers to that!

Yet something that I feel it lacks is that in the current implementation where it's using the `PUT` method we need to pass an existing `UUID` for a project to them implement the task. And getting this id requires that anyone using the solution need to manually go to their `DataTrack` portal to grab it, our to implement prior steps in the pipeline to get this info through the `API`.

I think a cleaver and most autonomous implementation would be to use the same route yet with a `POST` request and make available to users an option to turn on/off the `autoCreate` property, which would enable pipelines to upload a bill of material to a data track on an existing project or to create a new one if it doesn't exist.

Something in these lines:

```cmd
curl -X "POST" "https://dtrack-url.com/api/v1/bom" \
        -H 'X-Api-Key: $(api-key)' \
        -F '$(Build.Repository.Name)' \
        -F 'autoCreate=true' \
        -F 'projectVersion=latest' \
        -F 'bom=@$(Agent.TempDirectory)/bom.xml'
```
I'm up to develop this new implementation, what do you guys think about it?

**Comment by Zargath** [comment]
Hi rashmitt,

I'm happy to hear you like the task! I'm for sure open to contributions :) This is actually something I've been wanting to implement, but haven't had time.

I'm wondering why you would change the request to a POST since the PUT offers the same auto create functionality. 

> Expects CycloneDX or SPDX (text or RDF) along and a valid project UUID. If a UUID is not specified, than the projectName and projectVersion must be specified. Optionally, if autoCreate is specified and ‘true’ and the project does not exist, the project will be created. In this scenario, the principal making the request will additionally need the PORTFOLIO_MANAGEMENT or PROJECT_CREATION_UPLOAD permission.

```
PUT http://dtrack-url.com/api/v1/bom
{
    "project": "string",
    "projectName": "string",
    "projectVersion": "string",
    "autoCreate": true,
    "bom": "string"
}
```

Thanks!

**Comment by raschmitt** [comment]
> I'm happy to hear you like the task! I'm for sure open to contributions :) This is actually something I've been wanting to implement, but haven't had time.
>

I'm more than happy to help.

> I'm wondering why you would change the request to a POST since the PUT offers the same auto create functionality.

My bad, currently I am using the route `v1/project` to get/create the uuid for an existing/new project to them use it on the task, so then I have to use `PUT`/`POST`. 

But you are right, we don't actually need to change the method, just to expose `autoCreate` and substitute the `dtrackProjId` for `projectName` and `projectVersion`.

I will work on a PR.

**Comment by Zargath** [comment]
Awesome! I look forward to working through it with you :)

**Comment by raschmitt** [comment]
I started coding the proposed implementation today, and then realized that by the current `API` provided by `Dependency Track` we wouldn't be able to `getProjectMetrics` without having the projects `uuid`.

So I started digging around and realized we could get it from `getProjectInfo`, and use it to `getProjectMetricsAsync` down the road.

Here is, a very simplified, blueprint, of my final solution for enabling the use of `autoCreate`, what do you think about it? 

![Diagrama em branco](https://user-images.githubusercontent.com/11604696/106540033-e35cec80-64dd-11eb-9cfe-66d612f3dffe.png)

**Comment by Zargath** [comment]
That makes a lot of sense to me. I remember seeing that issue in a few of the endpoints :)

**Comment by Zargath** [comment]
This has been released with version 1.15.0. Thank you!
